### DISCUSSION TITLE
Upgrading from 0.25.24 t0 0.27.8 - issues with Type.Intersect

### DISCUSSION THREAD

**Discussion by tgilino** [sinclairzx81/typebox#393] [Q&A] [answered]
We recently found the need to update to 0.26+ due to needing `Type.BigInt` however we are now facing issues related to (I believe) the breaking change documented in the [0.26.0 changelog about Intersect](https://github.com/sinclairzx81/typebox/blob/master/changelog/0.26.0.md#Intersect-Schema-Representation). It states that we can use `Type.Composite` for backwards compatibility, but after updating (compilation is ok), we are getting runtime errors of the form
```
TypeExtends: Unknown left type operand 'Unsafe'
```
An example of what we were previously using in 0.25.24:
```
Type.Union([
	Type.Object({
		A: Type.String({ description: 'A' }),
	}),
	Type.Object({
		B: Type.String({ description: 'B' }),
	}),
	Type.Pick(
		Type.Intersect([
			ATypeWithALotOfProperties,
			Type.Object({
				C: Type.Optional(CTypeEnumSchema),
			}),
		]),
		[A List of Properties ],
	),
])
```
Kind of at a loss where to go from here. I also saw in the examples section an `IntersectLegacy` but I wasn't able to figure out how to get it recognized by type builder. Any help would be very much appreciated!

**Comment by sinclairzx81** [comment, marked as answer]
@tgilino Hi, 

Yeah, there has been quite a few changes from 0.26.0 with respect to intersection types. The composite type is provided as a fallback, however to keep composite inline with the new type model, this type also performs an `X extends Y` per property check to determine if overlapping properties would yield `TNever`. This check is only supported for known types (not unsafe types, hence the `Extends` error you see). I may investigate ways to relax this check in future revisions.

### IntersectLegacy Type

You can just import the intersect example from [here](https://github.com/sinclairzx81/typebox/blob/master/example/legacy/intersect.ts) and replace `Type.Intersect` for `IntersectLegacy`

```typescript
import { IntersectLegacy } from './intersect-legacy' // you will need to copy and paste file into your project

// ...

Type.Pick(
  IntersectLegacy([  // Just replace instances of Type.Intersect with IntersectLegacy 
    TypeWithALotOfProperties,
    Type.Object({
      C: Type.Optional(CTypeEnumSchema),
    }),
  ]),
  [A List of Properties ],
),

```

### TypeBuilder with IntersectLegacy Type

If you want to include the legacy intersect type on the type builder, you will need to extend and re export the `ExtendedTypeBuilder`

```typescript
import { ExtendedTypeBuilder, ObjectOptions, TObject } from '@sinclair/typebox'
import { IntersectLegacy, TIntersectLegacy } from './legacy/intersect'

export class ApplicationTypeBuilder extends ExtendedTypeBuilder {
  public IntersectLegacy<T extends TObject[]>(objects: [...T], options: ObjectOptions = {}): TIntersectLegacy<T> {
    return IntersectLegacy<T>(objects, options)
  }
}
// re-export the new type builder
export const Type = new ApplicationTypeBuilder()
```
Then import the extended application type builder and use as normal.
```typescript
import { Type } from './application-types'

// ...

Type.Pick(
  Type.IntersectLegacy([  // Now available on TypeBuilder
    TypeWithALotOfProperties,
    Type.Object({
      C: Type.Optional(CTypeEnumSchema),
    }),
  ]),
  [A List of Properties ],
),
```
Hope this helps
S

**Reply by tgilino** [reply]
Thanks so much for the prompt reply! Was able to get it working exactly as described.
